pwiz_tools/Skyline/Model/TransitionGroup.cs

```diff
@@ -726,14 +726,20 @@ protected override IList<ExplicitLoss> CreateState(ExplicitLoss[] mods)
                 // there should be only one loss
                 TransitionLosses firstLosses = null;
                 List<TransitionLosses> allLosses = null;
-                HashSet<double> allLossMasses = null;
+                HashSet<Tuple<double, int>> allLossMassCharges = null;
                 foreach (var losses in potentialLosses)
                 {
-                    double lossMass = CalcTransitionLossesMass(type, cleavageOffset, massType, losses);
+                    double lossMass = CalcTransitionLossesMass(type, cleavageOffset, massType, losses, out var lossCharge);

```

Comment:
Would it be possible to get rid of this function call, and just call "CalcTransitionLosses" on line 744?
Then you would have a TransitionLosses object and you could take advantage of its ".Equals" implementation, and put it in the hashset, instead of having a HashSet&lt;Tuple&lt;double, int>>.

Then, you could also make "allLosses" into a "HashSet&lt;TransitionLosses>" and you would not need "allLossMassCharges".

Reply:
Unfortunately, we can't rely on the loss equality for the reason I explained above. Losses with unequal charges can end up colliding with differently charged losses in fragments where the fragmentation event causes a charge loss from the precursor that was not related to the loss.

Reply:
I improved the comment about why this only changes for precursor ions. It was a surprise to me too, and another confirmation of the importance of writing tests.